keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,497 @@
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Expose new metrics
+
+The scheduler will expose an optional set of metrics series that capture the resources a pod is considered to be "consuming" from a scheduling perspective at the current time.
+
+The series reported will be consistent with `kube_node_status_allocatable` to simplify queries that compare used capacity and available capacity.
+
+* `kube_pod_resource_requests` contains series for the consumed requested resources for a pod at a given time.
+  * The value is the quantity of the limit in `spec.containers.resources.requests`
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `resource` - the name of the resource as described in `spec.containers.resources.requests`
+    * `units` - the units of the resource type as inferred from the resource name in `spec.containers.requests`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+* `kube_pod_resource_limits` contains series for the consumed limit resources for a pod at a given time.
+  * The value is the quantity of the limit in `spec.containers.resources.limits`
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `resource` - the name of the resource as described in `spec.containers.resources.limits`
+    * `units` - the units of the resource type as inferred from the resource name in `spec.containers.limits`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+* `kube_pod_resource_requests_lifecycle` contains series for the consumed requested resources for a pod at a given time when the pod is in the provided lifecycle phase. All pods are in one of the lifecycle phases, but series will only be returned for the current phase.
+  * The value is the quantity of the limit in `spec.containers.resources.requests`
+  * It has the following labels:
+    * `lifecycle` - this is set to the current lifecycle phase as described in the resource model - `Pending`, `Initializing`, `Running`, `Completed`
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `resource` - the name of the resource as described in `spec.containers.resources.requests`
+    * `units` - the units of the resource type as inferred from the resource name in `spec.containers.requests`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+* `kube_pod_resource_limits_lifecycle` contains series for the consumed limit resources for a pod at a given time when the pod is in the provided lifecycle phase.
```

Comment:
I don't see any motivation that would require having this metric for limits.  Having the `_lifecycle` metrics seems to be mostly about answering questions about resources a pod is allocated, which only requires knowing requests.

Reply:
Limits go into cgroup calculations too.  As a user of the system, requests and limits are imposed by different systems but I wouldn't view them as completely distinct.  if there is an effective limit during initializing but not during runtime and runtime shows evidence of hitting the limit, I should potentially adjust my limits to be higher for runtime.  I don't think it's out of bounds to in the future have additional subsystem interactions that consider limit where today it is ignored for some resources - for instance network bandwidth might be actively shaped.

Reply:
I'm mostly just suggesting we add something to the motivation or user journeys to talk about what problems we can catch by using this metric.
Although, i'm not sure I understand the scenario you outlined above.  Generally, reasoning about limits WRT throttling can be done entirely at the container level, as pod limits are always higher (i.e. less restrictive) that the sum of container limits.  My understanding is that pod limits matter in two scenarios:
1. When using container runtimes that have significant PodOverhead (e.g. kata), as that counts against the pod limit, but not individual container limits
2. When quotas are enforced on limits.  I think you may be assuming this, or something like this.  For example, if the pod limit is higher because of your init containers, you would be "wasting" quota.